metsim/metsim.py

```diff
@@ -42,71 +47,143 @@ def __init__(self, params:dict):
+                               MetSim.params, disagg=True))
+        queue.put(out_dict)
 
 
     def find_elevation(self, lat: float, lon: float) -> float:
         """ Use the domain file to get the elevation """
-        lat_idx = np.abs(self.lats - lat).argmin()
-        lon_idx = np.abs(self.lons - lon).argmin()
-        return self.elev[lat_idx, lon_idx]
+        lat_idx = np.abs(self.domain_lats - lat).argmin()
+        lon_idx = np.abs(self.domain_lons - lon).argmin()
+        return self.domain['elev'][lat_idx, lon_idx]
 
 
     def update(self, new_params):
         """Updates the global parameters dictionary"""
         MetSim.params.update(new_params)
 
 
-    def write(self, data: dict):
-        """
-        Dispatch to the right function based on the configuration given
-        """
-        dispatch = {
-                'netcdf' : self.write_netcdf,
-                'ascii'  : self.write_ascii
-                }
-        dispatch[MetSim.params.get('out_format', 'netcdf').lower()](data)
+    def netdf_in_preprocess(self, filename):
+        """Get the extent and load the data"""
+        # Get the information for splitting up the job
+        in_forcing = Dataset(filename, 'r')
+        self.locations = list(itertools.product(in_forcing['lat'], in_forcing['lon']))
+        self.lats = np.array(in_forcing['lat'])
+        self.lons = np.array(in_forcing['lon'])
+        self.lat_idx = {str(lat): i for i, lat in enumerate(self.lats)}
+        self.lon_idx = {str(lon): j for j, lon in enumerate(self.lons)}
+        in_forcing.close()
+        # Now read in the file properly
+        met_data = self.read(filename)
+        return met_data
 
 
-    def init_netcdf(self):
+    def ascii_in_preprocess(self, job_list):
+        """Process all files to find spatial extent"""
+        # Binary forcing files have naming format $NAME_$LAT_$LON
+        sets = [os.path.basename(fpath).split("_") for fpath in job_list]
+        self.lats = np.unique(sorted([float(s[1]) for s in sets]))
+        self.lons = np.unique(sorted([float(s[2]) for s in sets]))
+        self.lat_idx = {str(lat): i for i, lat in enumerate(self.lats)}
+        self.lon_idx = {str(lon): j for j, lon in enumerate(self.lons)}
+
+
+    def binary_in_preprocess(self, job_list):
+        """Process all files to find spatial extent"""
+        # Binary forcing files have naming format $NAME_$LAT_$LON
+        sets = [os.path.basename(fpath).split("_") for fpath in job_list]
+        self.locations = [(float(s[1]), float(s[2])) for s in sets]
+        self.lats = np.unique(sorted([float(s[1]) for s in sets]))
+        self.lons = np.unique(sorted([float(s[2]) for s in sets]))
+        self.lat_idx = {str(lat): i for i, lat in enumerate(self.lats)}
+        self.lon_idx = {str(lon): j for j, lon in enumerate(self.lons)}
+
+        # Creates the master dataset which will be used to parallelize
+        met_data = xr.Dataset(coords={'time' : MetSim.params['dates'],
```

Comment:
No need to have a master dataset that covers the entire domain. You only need to keep track of active cells and process those. Then inactive ones you can handle as part of the write statement only. This will speed up the process and reduce memory requirements.